### ISSUE TITLE
Feature request: Add Assert::ok($message) as contrast to Assert::fail($message)

### ISSUE THREAD

**Issue by JanTvrdik** [nette/tester#250]
https://forum.nette.org/cs/24204-jak-testovat-ze-neco-nehodi-vyjimku?

**Comment by dg** [comment]
This is IMHO better

```
	/**
	 * @param  callable
	 * @return void
	 */
	public static function noError($function)
	{
		self::error($function, array());
	}
```

**Comment by richard-ejem** [comment]
@dg yes that is simple and nice :) but this covers only errors, not thrown exceptions. I'd still add support for entering nice assertion message to let testers describe what should have happenned but was interrupted by an error, like this:

	/**
	 * Checks if the function does neither generate PHP error nor throw Exception
	 * @param  callable
	 * @param  string
	 * @return null|\Exception
	 */
	public static function noError($function, $message = NULL)
	{
		self::$counter++;

		set_error_handler(function ($severity, $msg, $file, $line) use (& $expected, $message) {
			if (($severity & error_reporting()) !== $severity) {
				return;
			}
			$errorStr = Helpers::errorTypeToString($severity) . ($msg ? " ($msg)" : '');
			restore_error_handler();
			Assert::fail(($message ? "$message: " : '') . "occurred error $errorStr");
		});
		try {
			call_user_func($function);
		} catch (\Exception $e) {
			restore_error_handler();
			Assert::fail(($message ? "$message: " : '') . sprintf('%s was thrown: %s (%d)',
				get_class($e), $e->getMessage(), $e->getCode()));
		}
		restore_error_handler();
	}

**Comment by milo** [comment]
@richard-ejem The `Assert::error()` is sufficient. It will be consistent across all assertions, unexpected exception will be just thrown. In your case, stack trace is missing.

**Comment by Mikulas** [comment]
:+1: for `Assert::ok`. Anything longer defeats the purpose as we already can write `Assert::true(TRUE, $msg)`.

**Comment by richard-ejem** [comment]
Assert::noError is maybe more expressive, but Assert::ok is also ok :)
@milo this is all about clear semantics - for me it's ugly to use Assert::error for expecting no error. It is for cases when I test a piece of code for not throwing particular error/exception (like the example in which my method does not throw InvalidArgumentException on argument types that should be accepted), which is semantically different case than a random unexpected error breaking my test.

Maybe also adding type of error/exception that is explicitly expected to NOT occur can be also adder, like Assert::noError(function() { if (false) throw new Nette\InvalidStateException; }, 'Nette\InvalidStateException', 'false is false also today'); - will provide clear test failure message when this exception occurs and full stack trace if anything other screws up.

**Comment by milo** [comment]
@richard-ejem I don't force you to use `Assert::error()`, I wrote about `Asset::noError()` implementaion detail.

**Comment by richard-ejem** [comment]
@milo sorry, I didn't get that. Anyway, I still think my solution covers it in a cleaner way.

**Comment by richard-ejem** [comment]
Ok, thanks for basically copying my idea into your code ;)

**Comment by milo** [comment]
@richard-ejem I really don't want to steal anyone's ideas or code. The solution is based on https://forum.nette.org/cs/24204-jak-testovat-ze-neco-nehodi-vyjimku#p162036 and I supposed when I opened PR #251 before yours #252 it's OK.

But if you think you should be the commit author, let me know, I'll force push it again with your authority.

**Comment by richard-ejem** [comment]
Oh sorry, I meant to write to [#254], I just opened wrong window There you led me to fix several things and then closed the issue with your pullrequest doing basically the same. It was just a bit surprise, but never mind,don't force-push anything because of that ;)

**Comment by milo** [comment]
Well, hard to explain. I know Tester's code pretty well and solving the #254 by error handler was a logical way. The #257 wasn't get well, so I came back to your PR. I proposed some changes but it is very hard and time consuming to navigate the proposer. Almost every time is much faster (and thats important to me) and easier for me write a failing test and just fix the issue.

Anyway, sorry about that feeling, I'll me more careful last time.
